### PR TITLE
Skaffold for Fission

### PR DESCRIPTION
**Pull request by vishal-biyani**
This PR introduces [multi stage Dockerfile](https://docs.docker.com/develop/develop-images/multistage-build/) for building various components and [Skaffold](https://skaffold.dev/) for development workflow.

I am not entirely convinced of the benefits of Skaffold for development workflow yet and still doing some experimentation. If I am not satisfied then will pull out Skaffold and only retain docker multi-stage builds.

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1172)

### CODE REVIEW

File: skaffold.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+####################################
```
Comment (life1347) [inline skaffold.yaml]: Change it a bit that should work
```
apiVersion: skaffold/v1beta9
kind: Config
build:
  artifacts:
    - image: <DOCKRHUB_REPO>/fission
      context: .
      docker:
        dockerfile: fission-bundle/Dockerfile.fission-bundle
    - image: <DOCKRHUB_REPO>/fetcher
      docker:
        dockerfile: environments/fetcher/cmd/Dockerfile.fission-fetcher
    - image: <DOCKRHUB_REPO>/preupgradechecks
      docker:
        dockerfile: preupgradechecks/Dockerfile.fission-preupgradechecks
  tagPolicy:
    envTemplate:
      template: "{{.IMAGE_NAME}}:skaffold-test"
deploy:
  helm:
    releases:
      - name: fission
        chartPath: ./charts/fission-all
        valuesFiles:
          - ./charts/fission-all/values.yaml
        namespace: "fission"
        setValues:
          fetcherImage: <DOCKRHUB_REPO>/fetcher
          image: <DOCKRHUB_REPO>/fission
          preUpgradeChecksImage: <DOCKRHUB_REPO>/preupgradechecks
          imageTag: skaffold-test
          # Set to "index.docker.io" if you're using local registry
          repository: index.docker.io
        setValueTemplates: {}
        wait: true
        recreatePods: false
        overrides: {}
        packaged: null
        imageStrategy:
          fqn: null
          helm: null
```

File: skaffold.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        dockerfile: Dockerfile.fission-bundle
+    - image: <DOCKRHUB_REPO>/fetcher
+      docker:
+        dockerfile: environments/fetcher/cmd/Dockerfile.fission-fetcher
+    - image: <DOCKRHUB_REPO>/preupgradechecks
+      docker:
+        dockerfile: preupgradechecks/Dockerfile.fission-preupgradechecks                
+  tagPolicy:
+    envTemplate:
+      template: "{{.IMAGE_NAME}}:{{.TAG}}"
```
Comment (life1347) [inline skaffold.yaml]:
```
  tagPolicy:
    envTemplate:
      template: "{{.IMAGE_NAME}}:skaffold-test"
```

**Comment by soamvasani** [comment]
Adding to 1.3 since it seems pretty far along.  Even if we don't end up using skaffold, the multi-stage build stuff is still very useful

File: skaffold.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+deploy:
+  helm:
+    releases:
+      - name: fission
+        chartPath: ./charts/fission-all
+        valuesFiles:
+          - ./charts/fission-all/values.yaml
+        namespace: "fission"
+        setValues:
+          fetcherImage: <DOCKRHUB_REPO>/fetcher
```
[the remark this answers is not in the document]
Comment (vishal-biyani) [reply skaffold.yaml]: This is intentional @life1347 - every developer will use their own Github repo and will have to replace `<DOCKERHUB_REPO>` with whatever their user handle is. Or did I misunderstood your comment?

[the remark this answers is not in the document]
Comment (life1347) [reply skaffold.yaml]: Yes, it works well if using dockerhub, but what if user want to pull from other registries like gcr.io ?

**Comment by vishal-biyani** [comment]
@life1347 I have addressed most comments on this PR, please take a look. I will follow up on build getting green (Right now it is hitting 50 minutes)

File: skaffold.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# 1) You can not refer to dynamically generated tag from build section in Helm values
+# - hence you have to use "latest" tag
+############### Possible enhancements ##############
+# 1) Add profiles to suit various deployment needs
+####################################
+apiVersion: skaffold/v1beta9
+kind: Config
+build:
+  artifacts:
+    - image: vishalbiyani/fission
```
Comment (life1347) [inline skaffold.yaml]: Maybe you want to replace `vishalbiyani` with `<DOCKERHUB_REPO>`?

File: skaffold.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+####################################
+# This file can be used with Skaffold (https://github.com/GoogleContainerTools/skaffold) to
+# build and deploy Fission to Kubernetes cluster.
+#
+############## Usage ##############
+# Skaffold CLI should be installed on your machine. 
+# Change vishalbiyani -> to your Dockerhub username
+# RUn `go mod vendor` so that all dependencies are downloaded
```
Comment (life1347) [inline skaffold.yaml]: `RUn` -> `Run`

File: test/test_utils.sh [hunk 6]
```diff
@@ -54,57 +54,59 @@ gcloud_login() {
...
+    echo $(git rev-parse HEAD)
+}
+
 build_and_push_pre_upgrade_check_image() {
     image_tag=$1
 
-    pushd $ROOT/preupgradechecks
-    ./build.sh
-    docker build -q -t $image_tag .
+    #pushd $ROOT/preupgradechecks
```
Comment (life1347) [inline test/test_utils.sh]: Remove comments if no longer needed

**Pull request merged**